### DISCUSSION TITLE
Migrate old contact data to new contacts model

### DISCUSSION THREAD

**Discussion by dejantep** [netbox-community/netbox#8169] [Q&A]
Hello,

Is there a way to do it more less automatically? I will migrate data from netbox 2.11.12 to 3,1,2
Should i migrate first then do export of sites, extract contact data, modify and then import to new contacts model?

**Comment by candlerb** [comment]
You haven't said where your contact data stored currently - in custom fields, in description, somewhere else?

In any case, you're not going to lose anything by moving to 3.1.2, so I'd suggest you do that migration first.  Then you'll have all the new objects and fields you need, and you can migrate data in whatever way makes sense.

(As usual, there are API changes when moving between versions, so if you have any existing REST API integrations, custom scripts, plugins etc, then you'll also need to make sure those are updated to work with 3.1)

**Reply by kkthxbye-code** [reply]
I mean he mentions sites and sites have contact fields that are going to be removed in the future, so it's probably from there (contact_name, contact_phone, contact_email, physical_address, shipping_address). These fields match the contact model.

**Reply by dejantep** [reply]
Hello guys!

Yes, im referring to information within sites. I already tried to export via gui but default export doesn't contain this information.

**Reply by candlerb** [reply]
Even "Export > All Data"?  If not, then you can make a custom export template.

I've just checked a 3.1 which was migrated.  There is a Site which still shows Contact Phone and Contact E-mail attributes.  Next to them is a yellow warning triangle with tooltip:

> This field will be removed in a future release. Please migrate this data to contact objects.

So the data is still there.  Physical Address and Shipping Address do not have this flag, so I think those are going to remain as-is.  (The address of the site itself is unrelated to any individual contacts at that site)

I've also checked Export > All Data, and that *does* include the contact info.

**Reply by candlerb** [reply]
Just checked the changelogs for future 3.2:

> The `contact_name`, `contact_phone`, and `contact_email` fields have been removed from the site model. Please use the new contact model introduced in NetBox v3.1 to store contact information for sites.

So you'd better migrate to 3.1 now, while your existing contact data is retained, before migrating to 3.2 in future.

**Comment by DanSheps** [comment]
Give me a couple of days and I will write up a script for this and post it here

**Reply by dejantep** [reply]
That would be fantastic. 👍

**Reply by DanSheps** [reply]
Haven't forgot about this, just busy

**Reply by DanSheps** [reply]
This is untested but... This should do it:

```python
from dcim.models import Site
from tenancy.models import Contact, ContactAssignment, ContactRole
from extras.scripts import Script, FileVar, StringVar, ObjectVar, TextVar


class UpdateContacts(Script):
    name = 'Update Contacts'
    description = 'Transition from contacts on object to new contact users'
    
    role = ObjectVar(
        label='Default Contact Role',
        description='Select the role to assign to any contacts created or assigned to a site',
        model=ContactRole,
        required=True
    )
    
    def run(self, data, commit):
        sites = Site.objects.all()
        for site in sites:
            if site.contact_name is not None:
                try:
                    contact = Contact.objects.get(name=site.contact_name)
                    self.log_info(f'For {site}; found matching contact {contact}')
                except Contact.MultipleObjectsFound:
                    # You could get deeper by trying to match on phone and email then trying no group
                    try:
                        contact = Contact.objects.get(name=site.contact_name, group=None)
                        self.log_info(f'For {site}; found matching contact {contact}')
                    except Contact.DoesNotExist:
                        contact = Contact.objects.filter(name=site.contact_name).first()
                        self.log_info(f'For {site}; found multiple matching contacts, using first contact {contact}')
                except Contact.DoesNotExist:
                    contact = Contact.objects.create(name=site.contact_name, phone=site.contact_phone, email=site.contact_email)
                    self.log_info(f'For {site}; created contact {contact}')
                
                contact_assignment = ContactAssignment.objects.create(object=site, contact=contact, role=self.role)
                self.log_info(f'Contact Assignment created for site: {site} and contact: {contact}')

```

**Reply by candlerb** [reply]
Debugged version:

```python
from dcim.models import Site
from tenancy.models import Contact, ContactAssignment, ContactRole
from extras.scripts import Script, FileVar, StringVar, ObjectVar, TextVar


class UpdateContacts(Script):
    name = 'Update Contacts'
    description = 'Transition from contacts on object to new contact users'

    role = ObjectVar(
        label='Default Contact Role',
        description='Select the role to assign to any contacts created or assigned to a site',
        model=ContactRole,
        required=True
    )

    def run(self, data, commit):
        sites = Site.objects.all()
        for site in sites:
            if site.contact_name:
                try:
                    contact = Contact.objects.get(name=site.contact_name)
                    self.log_info(f'For {site}; found matching contact {contact}')
                except Contact.MultipleObjectsReturned:
                    # You could get deeper by trying to match on phone and email then trying no group
                    try:
                        contact = Contact.objects.get(name=site.contact_name, group=None)
                        self.log_info(f'For {site}; found matching contact {contact}')
                    except Contact.DoesNotExist:
                        contact = Contact.objects.filter(name=site.contact_name).first()
                        self.log_info(f'For {site}; found multiple matching contacts, using first contact {contact}')
                except Contact.DoesNotExist:
                    contact = Contact.objects.create(name=site.contact_name, phone=site.contact_phone, email=site.contact_email)
                    self.log_info(f'For {site}; created contact {contact}')

                contact_assignment = ContactAssignment.objects.create(object=site, contact=contact, role=data["role"])
                self.log_info(f'Contact Assignment created for site: {site} and contact: {contact}')
```

**Reply by DanSheps** [reply]
Thanks, been a busy day

**Reply by candlerb** [reply]
No, thank you for writing it.  I had a handful of site contacts that needed migrating, and this sorted it out for me :-)

**Reply by candlerb** [reply]
One thing I noticed with the first version of the script: it was happy to create contacts with an empty name.  I'm not sure if that's intentional in the data model, although I suppose it's reasonable (you might only have a phone number, and might not want to put a fake name in)

That's why I changed to `if site.contact_name:` (since the name can never be null, but it can be empty string).

I do however note that if a site has a phone number or E-mail *without* a contact name, then your script won't migrate it.

**Reply by candlerb** [reply]
I found I did have some site contacts without name, which weren't migrated.  I modified the script again:

```python
from dcim.models import Site
from tenancy.models import Contact, ContactAssignment, ContactRole
from extras.scripts import Script, FileVar, StringVar, ObjectVar, TextVar
from django.contrib.contenttypes.models import ContentType


class UpdateContacts(Script):
    name = 'Update Contacts'
    description = 'Transition from contacts on object to new contact users'

    role = ObjectVar(
        label='Default Contact Role',
        description='Select the role to assign to any contacts created or assigned to a site',
        model=ContactRole,
        required=True
    )

    def run(self, data, commit):
        sites = Site.objects.all()
        site_content_type = ContentType.objects.get_for_model(Site)
        for site in sites:
            if site.contact_name or site.contact_phone or site.contact_email:
                contact = Contact.objects.filter(name=site.contact_name or "Unknown", phone=site.contact_phone, email=site.contact_email).first()
                if contact:
                    self.log_info(f'For {site}; found matching contact {contact.name} {contact.phone} {contact.email}')
                    if ContactAssignment.objects.filter(contact=contact, content_type=site_content_type, object_id=site.id).exists():
                        self.log_info(f'Contact assignment already exists, skipping')
                        continue
                else:
                    contact = Contact.objects.create(name=site.contact_name or "Unknown", phone=site.contact_phone, email=site.contact_email)
                    contact.full_clean()
                    self.log_info(f'For {site}; created contact {contact.name} {contact.phone} {contact.email}')

                contact_assignment = ContactAssignment.objects.create(object=site, contact=contact, role=data["role"])
                self.log_info(f'Contact Assignment created for site: {site} and contact: {contact.name} {contact.phone} {contact.email}')
```

EDIT: also it wasn't doing validation, which is why I was able to create contacts with no name. Added `contact.full_clean()`. Thanks to @kkthxbye-code [here](https://github.com/netbox-community/netbox/issues/8308#issuecomment-1009890901).

**Reply by DanSheps** [reply]
Yeah, that was an oversight (not running full_clean) on my part.

**Reply by rkwl1** [reply]
I also noticed that I have a caution triangle next to my site contact phone numbers.   All of my sites have a site contact phone number.  I would like to transition this to contacts, and assign it to the appropriate site.  I have over 200 sites that I need to migrate.
My programing skills are weak.  I have a dev Netbox that I can test this script on.  Is this something I would just run on the netbox server itself?  Thanks in advance for any help

**Reply by candlerb** [reply]
The code I posted above is a [Netbox Custom Script](https://netbox.readthedocs.io/en/stable/customization/custom-scripts/). You just drop this file in the `/opt/netbox/netbox/scripts/` directory, go to the web UI, select Other > Integrations > Scripts, and click on the script to run it.

**Reply by rkwl1** [reply]
This is brilliant.  it worked first time in my dev setup.  what if I wanted to run the script and instead of "Unknown" could I modify the script to say "Main Line" or something to that effect?

**Reply by rkwl1** [reply]
I figured this out.  Thanks again for the script.

**Reply by F3d0s** [reply]
> I found I did have some site contacts without name, which weren't migrated. I modified the script again:
> 
> ```python
>python script
> ```
> 
> EDIT: also it wasn't doing validation, which is why I was able to create contacts with no name. Added `contact.full_clean()`. Thanks to @kkthxbye-code [here](https://github.com/netbox-community/netbox/issues/8308#issuecomment-1009890901).

This worked for me better than [netbox_v32_migration](https://github.com/netbox-community/migration-scripts), which for some reasons didn't imported the contacts to the new schema used from v3.2.
